### PR TITLE
Set explicit target directory to avoid cargo deadlock

### LINKED ISSUE #9360: Build script of `proc_macro_test` deadlocks
It tries to invoke cargo inside cargo, which deadlocks if they share the same `target` directory. I'm using a global `target` directory to share artifacts between different projects.
https://github.com/rust-analyzer/rust-analyzer/blob/1b05dbba39d5a4d46f321dc962df99038cddbf21/crates/proc_macro_test/build.rs#L20-L24

### PR DESCRIPTION
**Pull request by oxalica**
Fix #9360 

r? @jonas-schievink

### CODE REVIEW

File: crates/proc_macro_test/build.rs [hunk 1]
```diff
@@ -17,9 +17,20 @@ fn main() {
...
         .current_dir("imp")
-        .args(&["build", "-p", "proc_macro_test_impl", "--message-format", "json"])
+        .args(&[
+            "build",
+            "-p",
+            "proc_macro_test_impl",
+            "--message-format",
+            "json",
+            // Explicit specify the target directory to avoid using the same one which the parent
+            // cargo is using, or we'll deadlock.
```
Comment (bjorn3) [inline crates/proc_macro_test/build.rs:30]: *When an env var is used to for all cargo builds to use the same target dir.

**Comment by jonas-schievink** [comment]
This means we can probably use a shared workspace again, right?

File: crates/proc_macro_test/build.rs [hunk 2]
```diff
@@ -17,9 +17,22 @@ fn main() {
...
+            "proc_macro_test_impl",
+            "--message-format",
+            "json",
+            // Explicit override the target directory to avoid using the same one which the parent
+            // cargo is using, or we'll deadlock.
+            // This can happen when `CARGO_TARGET_DIR` is set or global config forces all cargo
+            // instance to use the same target directory.
+            "--target-dir",
+            &target_dir.to_str().unwrap(),
+        ])
```
Comment (jonas-schievink) [inline crates/proc_macro_test/build.rs:35]:
```suggestion
        ])
        .arg(&target_dir)
```

**Comment by oxalica** [comment]
> This means we can probably use a shared workspace again

Do you mean the global shared target dir? Yes, it now builds fine for me.

**Pull request merged**